File: leetcode/c++/LCR_176_ping_heng_er_cha_shu_lcof.cpp
// https://leetcode.cn/problems/ping-heng-er-cha-shu-lcof/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    bool res = true;
    int depth(TreeNode* root) {
        if (root == nullptr)
            return 0;
        int l = depth(root->left), r = depth(root->right);
        if (abs(l - r) > 1) {
            res = false;
            return 0;
        }
        return max(l, r) + 1;
    }
public:
    bool isBalanced(TreeNode* root) {
        if (root == nullptr)
            return true;
        depth(root);
        return res;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(n)

File: leetcode/golang/LCR_176_ping_heng_er_cha_shu_lcof.go
type TreeNode struct {
         Val int
         Left *TreeNode
         Right *TreeNode
}
   
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}

func isBalanced(root *TreeNode) bool {
    res := true
    var dfs func(root *TreeNode) int
    dfs = func(root *TreeNode) int {
        if root == nil || !res {
            return 0
        }
        l := dfs(root.Left)
        r := dfs(root.Right)
        if abs(l - r) > 1 {
            res = false
            return 0
        }
        return max(l, r) + 1
    }
    dfs(root)
    return res
}